stats/opentelemetry/internal/csm/pluginoption.go

```diff
@@ -0,0 +1,341 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package csm contains the implementation of the CSM Plugin Option.
+package csm
+
+import (
+	"context"
+	"encoding/base64"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/xds/bootstrap"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats/opentelemetry/internal"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+var logger = grpclog.Component("csm-observability-plugin")
+
+// AddLabels adds CSM labels to the provided context's metadata, as a encoded
+// protobuf Struct as the value of x-envoy-metadata.
+func (cpo *PluginOption) AddLabels(ctx context.Context) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, metadataExchangeKey, cpo.metadataExchangeLabelsEncoded)
+}
+
+// NewLabelsMD returns a metadata.MD with the CSM labels as a encoded protobuf
+// Struct as the value of x-envoy-metadata.
+func (cpo *PluginOption) NewLabelsMD() metadata.MD {
+	return metadata.New(map[string]string{
+		metadataExchangeKey: cpo.metadataExchangeLabelsEncoded,
+	})
+}
+
+// GetLabels gets the CSM peer labels from the metadata provided. It returns
+// "unknown" for labels not found. Labels returned depend on the remote type.
+// Additionally, local labels determined at initialization time are appended to
+// labels returned, in addition to the optionalLabels provided.
+func (cpo *PluginOption) GetLabels(md metadata.MD, optionalLabels map[string]string) map[string]string {
+	labels := map[string]string{ // Remote labels if type is unknown (i.e. unset or error processing x-envoy-peer-metadata)
+		"csm.remote_workload_type":              "unknown",
+		"csm.remote_workload_canonical_service": "unknown",
+	}
+	// Append the local labels.
+	for k, v := range cpo.localLabels {
+		labels[k] = v
+	}
+
+	// Append the optional labels. To avoid string comparisons, assume the
+	// caller only passes in two potential xDS Optional Labels: service_name and
+	// service_namespace.
```

Comment:
Previously, this layer was meant to "filter" or "eat" labels that weren't one of "service_name" or "service_namespace". I initially allowed all labels through CDS, however switched CDS to only process these two labels if set and if string value: https://github.com/grpc/grpc-go/commit/308dbc446644b2f783c29a7abc03f9bec60e8512#diff-976ef43c72aa5b948da568948f07eaf816e7e7d51704259e3ebafc1d7284d376R89. Way back in the o11y thread I had a back and forth with Yash about what to do in each layer. Eventually, Craig Yash and Mark decided to do this check in xDS Client rather than here to avoid costly string comparisons happening on the RPC path.